### PR TITLE
feat(EMS-2547-2675): No PDF - Buyer - Trading history and alternative currency - Data saving

### PR DESCRIPTION
**Pull request by Zainzzkk**
## Introduction :pencil2:
This PR adds validation for alternative currency and data saving for trading history and alternative currency

## Resolution :heavy_check_mark:
* Added to schema for buyerTradingHistory table
* Added saving, mapping in the ui for buyerTradingHistory
* Added validation for alternative trading address input
* Added population of fields from submitted data
* Added tests
* Updated e2e tests for data saving

## Miscellaneous :heavy_plus_sign:
* split save-data, map-submitted-data,  map-and-save in your-buyer section to buyer and buyer-trading-history folders
* removed get-country-by-name
* changed to use get-county-by-iso-code on country autocomplete

### CODE REVIEW

File: .cspell.json [hunk 1]
```diff
@@ -20,6 +20,7 @@
         "**logs",
         "**/src/api/test-mocks/mock-application.ts",
         "**/src/ui/server/test-mocks/mock-application.ts",
+        "**/src/ui/server/test-mocks/mock-buyer.ts",
```
Comment (ttbarnes) [inline .cspell.json:23]: Why is this required?

File: e2e-tests/commands/insurance/assert-alternative-currency-form.js [hunk 2]
```diff
@@ -9,9 +9,28 @@ import {
...
 
+/**
+ * returns field for currency/alternative currency
+ * if currency provided, then returns radio for respective currency option
+ * if no currency provided, then returns radio for alternative currency
+ * @param {String} FIELD_ID - fieldId for currency radio
+ * @param {String} ALTERNATIVE_CURRENCY_FIELD_ID - fieldId for alternative currency radio
+ * @param {String} currency - Currency string
+ */
+const whichRadio = (FIELD_ID, ALTERNATIVE_CURRENCY_FIELD_ID, currency) => {
```
Comment (ttbarnes) [inline e2e-tests/commands/insurance/assert-alternative-currency-form.js:24]: Can we rename `whichRadio`? Too generic, also can we have camel case params here? Otherwise look like constants

Also need documentation for `@return`

File: e2e-tests/commands/insurance/assert-alternative-currency-form.js [hunk 3]
```diff
@@ -21,22 +40,24 @@ const { CURRENCY: { CURRENCY_CODE, ALTERNATIVE_CURRENCY_CODE } } = INSURANCE_FIE
...
-    const { option: option1 } = radios(FIELD_ID, EUR.isoCode);
-    const { option: option2 } = radios(FIELD_ID, GBP.isoCode);
-    const { option: option3 } = radios(FIELD_ID, USD.isoCode);
-    const { option: option4 } = radios(FIELD_ID, JPY.isoCode);
-    const { option: option5 } = radios(ALTERNATIVE_CURRENCY_FIELD_ID);
+    const option1 = whichRadio(FIELD_ID, ALTERNATIVE_CURRENCY_FIELD_ID, EUR.isoCode);
+    const option2 = whichRadio(FIELD_ID, ALTERNATIVE_CURRENCY_FIELD_ID, GBP.isoCode);
+    const option3 = whichRadio(FIELD_ID, ALTERNATIVE_CURRENCY_FIELD_ID, USD.isoCode);
+    const option4 = whichRadio(FIELD_ID, ALTERNATIVE_CURRENCY_FIELD_ID, JPY.isoCode);
+    const option5 = whichRadio(FIELD_ID, ALTERNATIVE_CURRENCY_FIELD_ID);
```
Comment (ttbarnes) [inline e2e-tests/commands/insurance/assert-alternative-currency-form.js:60]: I think the original cleaner 🤔 Here we're passing `ALTERNATIVE_CURRENCY_FIELD_ID`  4 times for it not to be used?

File: e2e-tests/commands/insurance/assert-alternative-currency-form.js [hunk 4]
```diff
@@ -85,6 +106,73 @@ const assertAlternativeCurrencyForm = ({
...
+
+    cy.checkText(
+      partials.errorSummaryListItems().first(),
+      ERRORS[ALTERNATIVE_CURRENCY_FIELD_ID].IS_EMPTY,
+    );
+
+    cy.checkText(
+      field(ALTERNATIVE_CURRENCY_FIELD_ID).errorMessage(),
+      `Error: ${ERRORS[ALTERNATIVE_CURRENCY_FIELD_ID].IS_EMPTY}`,
+    );
```
Comment (ttbarnes) [inline e2e-tests/commands/insurance/assert-alternative-currency-form.js:123]: `cy.submitAndAssertFieldErrors`?

File: e2e-tests/commands/insurance/assert-alternative-currency-form.js [hunk 5]
```diff
@@ -85,6 +106,73 @@ const assertAlternativeCurrencyForm = ({
...
+  },
+  submitRadioRedirect: (CURRENCY, REDIRECT_URL) => {
```
Comment (ttbarnes) [inline e2e-tests/commands/insurance/assert-alternative-currency-form.js:125]: Not a fan of the uppercase params in this whole file, they look like constants


Function name is confusing can we rename? As it's selecting a radio, submitting and then asserting a redirect URL

File: e2e-tests/commands/insurance/assert-alternative-currency-form.js [hunk 6]
```diff
@@ -85,6 +106,73 @@ const assertAlternativeCurrencyForm = ({
...
+    const option = whichRadio(FIELD_ID, ALTERNATIVE_CURRENCY_FIELD_ID, CURRENCY);
+
+    option.input().click();
+    cy.clickSubmitButton();
+
+    cy.assertUrl(REDIRECT_URL);
+  },
+  submitRadioCheckRadio: (CURRENCY, url) => {
```
Comment (ttbarnes) [inline e2e-tests/commands/insurance/assert-alternative-currency-form.js:133]: `submitAndAssertRadioIsChecked`?

File: e2e-tests/commands/shared-commands/assertions/check-autocomplete-input.js [hunk 7]
```diff
@@ -56,6 +56,19 @@ const checkAutocompleteInput = {
...
   },
+  search: (field, input) => {
+    field.input().clear();
+
+    cy.keyboardInput(field.input(), input);
+    const results = field.results();
+
+    // select the first result
+    results.first().click();
+  },
```
Comment (ttbarnes) [inline e2e-tests/commands/shared-commands/assertions/check-autocomplete-input.js:67]: Once my PR is merged this can be replaced with `cy.autocompleteKeyboardInput`

File: e2e-tests/insurance/cypress/e2e/journeys/your-buyer/alternative-currency/alternative-currency-page.spec.js [hunk 8]
```diff
@@ -94,23 +99,77 @@ context('Insurance - Your Buyer - Alternative currency - As an exporter, I want
...
         cy.navigateToUrl(url);
       });
 
       it('should render validation errors', () => {
-        cy.clickSubmitButton();
+        rendersAlternativeCurrencyValidationError();
+      });
+    });
+
+    describe('when submitting a fully filmed form', () => {
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/your-buyer/alternative-currency/alternative-currency-page.spec.js:112]: filmed? Could we be more descriptive e.g "when submitting a non-alternative currency"

File: src/api/helpers/create-a-buyer-trading-history/index.test.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+  test('it should return a buyer trading history with respective IDs', async () => {
+    const result = await createABuyerTradingHistory(context, application.buyer.id);
+
+    expect(result).toBeDefined();
+    expect(typeof result.id).toEqual('string');
+    expect(result.id.length).toBeGreaterThan(0);
+  });
+
+  test('it should return empty buyerTradingAddress fields', async () => {
```
Comment (ttbarnes) [inline src/api/helpers/create-a-buyer-trading-history/index.test.ts:33]: ...with default `currencyCode`?

File: src/api/helpers/create-a-buyer-trading-history/index.ts [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+  try {
+    const buyerTradingHistory = await context.db.BuyerTradingHistory.createOne({
+      data: {
+        buyer: {
+          connect: {
+            id: buyerId,
+          },
+        },
+        currencyCode: GBP,
```
Comment (ttbarnes) [inline src/api/helpers/create-a-buyer-trading-history/index.ts:22]: Can we have this coming from `constants/application/versions`?

File: src/api/schema.ts [hunk 11]
```diff
@@ -533,10 +533,29 @@ export const lists = {
...
-      exporterHasTradedWithBuyer: nullableCheckbox(),
       connectionWithBuyerDescription: text({
         db: { nativeType: 'VarChar(1000)' },
       }),
+      buyerTradingHistory: relationship({ ref: 'BuyerTradingHistory.buyer' }),
+    },
+    hooks: {
+      afterOperation: async ({ item, context }) => {
+        if (item?.applicationId) {
+          await updateApplication.timestamp(context, item.applicationId);
```
Comment (ttbarnes) [inline src/api/schema.ts:544]: Can we update `keystone.test.ts` to include a test?

File: src/api/schema.ts [hunk 12]
```diff
@@ -533,10 +533,29 @@ export const lists = {
...
+        }
+      },
+    },
+    access: allowAll,
+  }),
+  BuyerTradingHistory: list({
+    fields: {
+      buyer: relationship({ ref: 'Buyer.buyerTradingHistory' }),
+      currencyCode: text({
+        db: { nativeType: 'VarChar(5)' },
```
Comment (ttbarnes) [inline src/api/schema.ts:554]: I think it'd be safe to has this as 3

File: src/ui/server/api/keystone/application/update/buyerTradingHistory.ts [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+import updateBuyerTradingHistoryMutation from '../../../../graphql/mutations/update-application/buyerTradingHistory';
+import { ApolloResponse } from '../../../../../types';
+
+/**
+ * updateBuyer
+ * Update an application's "buyer"
+ * @param {String} Buyer ID
+ * @param {Object} Buyer update
+ * @returns {Object} Updated buyer
+ */
```
Comment (ttbarnes) [inline src/ui/server/api/keystone/application/update/buyerTradingHistory.ts:11]: Should the documentation be for "buyer trading history" instead of "buyer"?

File: src/ui/server/constants/supported-currencies.ts [hunk 14]
```diff
@@ -4,3 +4,4 @@ export const EUR = 'EUR';
 export const GBP = 'GBP';
 export const JPY = 'JPY';
 export const USD = 'USD';
+export const AED = 'AED';
```
Comment (ttbarnes) [inline src/ui/server/constants/supported-currencies.ts:7]: Should this be in here? `AED` is not a standard/"supported" currency?

In my PR when testing for a non-standard currency - currently only for an E2E test, have added this to fixtures to be clearer:

```js
export const NON_STANDARD_CURRENCY_CODE = AED_CURRENCY_CODE;
```

File: e2e-tests/commands/insurance/assert-alternative-currency-form.js (hunk 4 above)
Comment (Zainzzkk) [reply e2e-tests/commands/insurance/assert-alternative-currency-form.js:123]: will not work because there are different fieldIds.  Everywhere where we have different fieldIds for the input needs it this way round

File: src/ui/server/controllers/insurance/your-buyer/alternative-currency/validation/rules/alternative-currency.ts [hunk 15]
```diff
@@ -1,10 +1,10 @@
 import INSURANCE_FIELD_IDS from '../../../../../../constants/field-ids/insurance';
 import { ERROR_MESSAGES } from '../../../../../../content-strings';
 import { RequestBody } from '../../../../../../../types';
-import emptyFieldValidation from '../../../../../../shared-validation/empty-field';
+import alternativeCurrencyValidation from '../../../../../../shared-validation/alternative-currency.ts';
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/alternative-currency/validation/rules/alternative-currency.ts:4]: No need for `.ts`

File: src/ui/server/controllers/insurance/your-buyer/alternative-currency/validation/rules/alternative-currency.test.ts [hunk 16]
```diff
@@ -1,16 +1,16 @@
 import alternativeCurrencyRule from './alternative-currency';
 import INSURANCE_FIELD_IDS from '../../../../../../constants/field-ids/insurance';
 import { ERROR_MESSAGES } from '../../../../../../content-strings';
-import emptyFieldValidation from '../../../../../../shared-validation/empty-field';
 import { RequestBody } from '../../../../../../../types';
+import alternativeCurrencyValidation from '../../../../../../shared-validation/alternative-currency.ts';
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/alternative-currency/validation/rules/alternative-currency.test.ts:5]: No need for `.ts`

File: src/ui/server/controllers/insurance/your-buyer/map-and-save/buyer-trading-history/index.test.ts [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+const {
+  CURRENCY: { CURRENCY_CODE, ALTERNATIVE_CURRENCY_CODE },
+  YOUR_BUYER: { FAILED_PAYMENTS },
+} = INSURANCE_FIELD_IDS;
+
+describe('controllers/insurance/your-buyer/map-and-save/buyer-trading-history', () => {
+  jest.mock('../../save-data/buyer-trading-history');
+
+  describe('buyer', () => {
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/map-and-save/buyer-trading-history/index.test.ts:17]: i think this can be removed, keep it simpler

File: src/ui/server/helpers/mappings/map-currencies/is-supported-currency/index.test.ts [hunk 18]
```diff
@@ -0,0 +1,20 @@
+import isSupportedCurrency from '.';
+import { GBP, AED } from '../../../../constants';
```
Comment (ttbarnes) [inline src/ui/server/helpers/mappings/map-currencies/is-supported-currency/index.test.ts:2]: Same as above, could we use `NON_STANDARD_CURRENCY_CODE` constant/fixture instead please so it's clearer?

File: src/ui/server/helpers/mappings/map-currencies/submitted-answer/index.ts [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+ * if currency is an alternative currency, then returns ALTERNATIVE_CURRENCY_CODE for radio to be selected
+ * @param {String} currency
+ * @returns {String} currency - empty string, currencyCode for supported currency or ALTERNATIVE_CURRENCY_CODE
+ */
+const submittedAnswer = (currency?: string) => {
+  if (!currency) {
+    return '';
+  }
+
+  if (isSupportedCurrency(currency)) {
```
Comment (ttbarnes) [inline src/ui/server/helpers/mappings/map-currencies/submitted-answer/index.ts:20]: Could we use the term "standard" instead of supported"?

I think it's clearer because another currency could be chosen, and it could potentially be supported?

What do you think?

File: e2e-tests/commands/insurance/assert-alternative-currency-form.js [hunk 20]
```diff
@@ -51,33 +74,98 @@ const assertAlternativeCurrencyForm = ({
...
-    checkAutocompleteInput.hasWorkingClientSideJS(countryInput.field(ALTERNATIVE_CURRENCY_FIELD_ID));
-    checkAutocompleteInput.rendersInput(countryInput.field(ALTERNATIVE_CURRENCY_FIELD_ID));
+    checkAutocompleteInput.hasWorkingClientSideJS(countryInput.field(alternativeCurrencyFieldId));
+    checkAutocompleteInput.rendersInput(countryInput.field(alternativeCurrencyFieldId));
   },
   doesNotRenderSupportedCurrencies: () => {
     cy.clickAlternativeCurrencyRadioOption();
 
-    checkAutocompleteInput.rendersNoResultsMessage(countryInput.field(ALTERNATIVE_CURRENCY_FIELD_ID), 'test');
+    checkAutocompleteInput.rendersNoResultsMessage(countryInput.field(alternativeCurrencyFieldId), 'test');
```
Comment (ttbarnes) [inline e2e-tests/commands/insurance/assert-alternative-currency-form.js:89]: `'test'` ?

File: e2e-tests/commands/shared-commands/assertions/check-autocomplete-input.js (hunk 7 above)
Comment (ttbarnes) [reply e2e-tests/commands/shared-commands/assertions/check-autocomplete-input.js:67]: Bump, this can be used/simplified now^

File: e2e-tests/insurance/cypress/e2e/journeys/your-buyer/alternative-currency/alternative-currency-page.spec.js [hunk 21]
```diff
@@ -97,23 +101,81 @@ context('Insurance - Your Buyer - Alternative currency - As an exporter, I want
...
+      });
+    });
+
+    describe('when submitting a supported currency', () => {
+      beforeEach(() => {
+        cy.navigateToUrl(url);
+      });
+
+      describe(EUR.name, () => {
+        const { isoCode } = EUR;
```
Comment (ttbarnes) [inline e2e-tests/insurance/cypress/e2e/journeys/your-buyer/alternative-currency/alternative-currency-page.spec.js:120]: For all the ISO codes in this test, could just import `EUR_CURRENCY_CODE`, `GBP_CURRENCY_CODE` etc

File: src/api/types/application-types/index.ts [hunk 22]
```diff
@@ -158,6 +165,7 @@ export interface ApplicationVersion {
   TOTAL_VALUE_OF_CONTRACT: number;
   DEFAULT_FINAL_DESTINATION_KNOWN: boolean;
   DEFAULT_NEED_PRE_CREDIT_PERIOD_COVER: boolean;
+  DEFAULT_CURRENCY?: string;
```
Comment (ttbarnes) [inline src/api/types/application-types/index.ts:168]: I think this should work without it being optional

File: src/ui/server/constants/currencies.ts [hunk 23]
```diff
@@ -1 +1,3 @@
 export const GBP_CURRENCY_CODE = 'GBP';
+
+export const NON_STANDARD_CURRENCY_CODE = 'AED';
```
Comment (ttbarnes) [inline src/ui/server/constants/currencies.ts:3]: Can we add a const for AED? In E2E we have this

```js
export const NON_STANDARD_CURRENCY_CODE = AED_CURRENCY_CODE;
```

File: src/ui/server/controllers/insurance/your-buyer/map-submitted-data/buyer-trading-history/index.test.ts [hunk 24]
```diff
@@ -0,0 +1,49 @@
...
+
+describe('controllers/insurance/your-buyer/map-submitted-data/buyer-trading-history', () => {
+  let mockFormBody: RequestBody;
+
+  beforeEach(() => {
+    mockFormBody = {
+      _csrf: 'abc',
+      [CURRENCY_CODE]: GBP,
+      [ALTERNATIVE_CURRENCY_CODE]: '',
+      other: 'test',
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/map-submitted-data/buyer-trading-history/index.test.ts:18]: Can we have something more descriptive than "test"

File: src/ui/server/helpers/mappings/map-currencies/is-not-alternative-currency/index.ts [hunk 25]
```diff
@@ -0,0 +1,12 @@
+import { SUPPORTED_CURRENCIES } from '../../../../constants';
+
+/**
+ * isSupportedCurrency
+ * checks if provided currency is provided or not
```
Comment (ttbarnes) [inline src/ui/server/helpers/mappings/map-currencies/is-not-alternative-currency/index.ts:5]: Typos 😛

File: src/ui/server/helpers/mappings/map-currencies/radio-and-select-options/index.ts [hunk 26]
```diff
@@ -14,12 +15,13 @@ const {
  * @param {Array} alternativeCurrencies: Array of alternative currencies
  * @param {Array} currencies: Array of supported currencies
  * @param {String} selectedValue: Selected currency
- * @returns {Object} Mapped currencies
+ * @returns {Object} Mapped currencies and applicationAnswer
```
Comment (ttbarnes) [inline src/ui/server/helpers/mappings/map-currencies/radio-and-select-options/index.ts:18]: Should this be `submittedValue`?

File: e2e-tests/insurance/cypress/e2e/journeys/your-buyer/alternative-currency/alternative-currency-page.spec.js (hunk 21 above)
Comment (Zainzzkk) [reply e2e-tests/insurance/cypress/e2e/journeys/your-buyer/alternative-currency/alternative-currency-page.spec.js:120]: I am using the name too though?

Comment (ttbarnes) [reply e2e-tests/insurance/cypress/e2e/journeys/your-buyer/alternative-currency/alternative-currency-page.spec.js:120]: Only for the test description though right?

Comment (Zainzzkk) [reply e2e-tests/insurance/cypress/e2e/journeys/your-buyer/alternative-currency/alternative-currency-page.spec.js:120]: yes, should I hard type the name like euros?

Comment (ttbarnes) [reply e2e-tests/insurance/cypress/e2e/journeys/your-buyer/alternative-currency/alternative-currency-page.spec.js:120]: I was thinking could just use the ISO code for all, e.g `EUR_CURRENCY_CODE`,  what do you think?

**Review by ttbarnes — APPROVED** [review]
